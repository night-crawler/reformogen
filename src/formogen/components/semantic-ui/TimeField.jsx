import React from 'react';
import { Form } from 'semantic-ui-react';
import moment from 'moment';
import _ from 'lodash';
import TimePicker from 'react-times';
import Label from './Label';

import propTypes from '../fieldPropTypes';


import 'react-times/css/material/default.css';


DateField.propTypes = propTypes;
export default function DateField(props) {
    let hour=null, minutes=null;

    const handleChange = (timeString) => {
        // trigger global onChange in case user've changed both minutes and hours
        // (calling onChange on every change causes redraw && closing TimePicker)
        hour && minutes && props.onChange(null, {
            name: props.name,
            // append seconds to time
            value: timeString ? moment(timeString, 'HH:mm').format('HH:mm:ss') : null
        });
    };

    let _props = {
        autoMode: true,
        name: props.name,
        time: props.value ? moment(props.value, 'HH:mm:ss').format('HH:mm') : null,
        timeMode: '24',
        focused: false,
        withoutIcon: false,
        showTimezone: true,
        onMinuteChange: (_minutes) => (minutes = _minutes),
        onHourChange: (_hour) => (hour = _hour),
        onTimeChange: handleChange,
        // timezone: 'America/New_York',
        theme: 'material',
    };

    if (_.isFunction(props.updateProps)) {
        _props = props.updateProps(_props, props);
    }

    return (
        <Form.Field required={ props.required } disabled={ !props.editable } width={ props.layoutOpts.width }>
            <Label { ...props } />
            <TimePicker { ..._props } />
            { !props.helpTextOnHover ? <span className="help-text">{ props.help_text }</span> : ''}
        </Form.Field>
    );
}
