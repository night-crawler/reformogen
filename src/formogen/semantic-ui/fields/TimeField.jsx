import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import moment from 'moment';
import { Form } from 'semantic-ui-react';
import TimePicker from 'react-times';

import { errorsType, layoutOptsType } from '../../fieldPropTypes';
import { FieldLabel } from '../FieldLabel';
import { ErrorsList } from '../ErrorsList';


export class TimeField extends React.Component {
  static propTypes = {
    type: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    value: PropTypes.string,
    max_length: PropTypes.number,
    help_text: PropTypes.string,
    errors: errorsType,
  
    required: PropTypes.bool,
    editable: PropTypes.bool,
  
    helpTextOnHover: PropTypes.bool,
    layoutOpts: layoutOptsType,
  
    updateProps: PropTypes.func,
    onChange: PropTypes.func,
  };

  constructor(props) {
    super(props);
    let [ hour, minute ] = [ null, null ];
    if (props.value) {
      [ hour, minute ] = moment(this.props.value, 'HH:mm:ss').format('HH:mm').split(':');
    }

    this.state = {
      hour, minute,
      meridiem: props.meridiem || null,
    };
  }

  
  render() {
    return (
      <Form.Field
        required={ this.props.required }
        disabled={ !this.props.editable }
        width={ this.props.layoutOpts.width }
        error={ !_.isEmpty(this.props.errors) }
      >
        <FieldLabel { ...this.props } />
        <TimePicker 
          autoMode={ true }
          timeMode='24'
          focused={ false }
          withoutIcon={ false }
          showTimezone={ true }
          onTimeChange={ this.handleTimeChange }
          onFocusChange={ this.handleFocusChange }
          theme='material'
          timezone={ this.props.timezone }

          time={ this.state.hour && this.state.minute 
            ? `${this.state.hour}:${this.state.minute}` 
            : null 
          }
          meridiem={ this.state.meridiem }
        />
        { !this.props.helpTextOnHover
          ? <span className='help-text'>{ this.props.help_text }</span>
          : '' 
        }
        <ErrorsList messages={ this.props.errors } />
      </Form.Field>
    );
  }

  handleTimeChange = options => {
    // eslint-disable-next-line
    this.setState(options);
  };

  handleFocusChange = focus => {
    // eslint-disable-next-line
    this.setState(
      { focus },
      () => !focus && this.props.onChange(null, { 
        name: this.props.name, 
        value: `${this.state.hour}:${this.state.minute}` 
      })
    );
  };
}
