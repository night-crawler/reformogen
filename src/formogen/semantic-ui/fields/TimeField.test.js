import { TimeField } from './TimeField';


it('<TimeField />', () => {
  const wrapper = shallow(
    <TimeField
      type='TimeField'
      name='TimeField: name'
      verbose_name='TimeField: verbose_name'
      help_text='TimeField: help_text'
      layoutOpts={ { width: 1 } }
    />
  );
  expect(wrapper).toMatchSnapshot();
});
