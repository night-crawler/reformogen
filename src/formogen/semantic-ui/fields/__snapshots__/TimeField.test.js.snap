// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeField /> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimeField
    displayOptions={
      Object {
        "width": 1,
      }
    }
    help_text="TimeField: help_text"
    name="TimeField: name"
    type="TimeField"
    verbose_name="TimeField: verbose_name"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <FieldLabel
          displayOptions={
            Object {
              "width": 1,
            }
          }
          help_text="TimeField: help_text"
          name="TimeField: name"
          type="TimeField"
          verbose_name="TimeField: verbose_name"
        />,
        <TimePicker
          autoClose={true}
          autoMode={true}
          closeOnOutsideClick={true}
          colorPalette="light"
          disabled={false}
          draggable={true}
          focused={false}
          language="en"
          limitDrag={false}
          meridiem={null}
          minuteStep={5}
          onFocusChange={[Function]}
          onTimeChange={[Function]}
          onTimezoneChange={[Function]}
          placeholder=""
          showTimezone={true}
          theme="material"
          time={null}
          timeConfig={
            Object {
              "step": 30,
              "unit": "minutes",
            }
          }
          timeFormat=""
          timeFormatter={null}
          timeMode="24"
          trigger={null}
          useTz={true}
          withoutIcon={false}
        />,
        <span
          className="help-text"
        >
          TimeField: help_text
        </span>,
        <ErrorsList
          color="red"
          header="Errors"
        />,
      ],
      "disabled": true,
      "error": false,
      "required": undefined,
      "width": 1,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "displayOptions": Object {
            "width": 1,
          },
          "help_text": "TimeField: help_text",
          "name": "TimeField: name",
          "type": "TimeField",
          "verbose_name": "TimeField: verbose_name",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoClose": true,
          "autoMode": true,
          "closeOnOutsideClick": true,
          "colorPalette": "light",
          "disabled": false,
          "draggable": true,
          "focused": false,
          "language": "en",
          "limitDrag": false,
          "meridiem": null,
          "minuteStep": 5,
          "onFocusChange": [Function],
          "onTimeChange": [Function],
          "onTimezoneChange": [Function],
          "placeholder": "",
          "showTimezone": true,
          "theme": "material",
          "time": null,
          "timeConfig": Object {
            "step": 30,
            "unit": "minutes",
          },
          "timeFormat": "",
          "timeFormatter": null,
          "timeMode": "24",
          "timezone": undefined,
          "trigger": null,
          "useTz": true,
          "withoutIcon": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "TimeField: help_text",
          "className": "help-text",
        },
        "ref": null,
        "rendered": "TimeField: help_text",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "color": "red",
          "header": "Errors",
          "messages": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <FieldLabel
            displayOptions={
              Object {
                "width": 1,
              }
            }
            help_text="TimeField: help_text"
            name="TimeField: name"
            type="TimeField"
            verbose_name="TimeField: verbose_name"
          />,
          <TimePicker
            autoClose={true}
            autoMode={true}
            closeOnOutsideClick={true}
            colorPalette="light"
            disabled={false}
            draggable={true}
            focused={false}
            language="en"
            limitDrag={false}
            meridiem={null}
            minuteStep={5}
            onFocusChange={[Function]}
            onTimeChange={[Function]}
            onTimezoneChange={[Function]}
            placeholder=""
            showTimezone={true}
            theme="material"
            time={null}
            timeConfig={
              Object {
                "step": 30,
                "unit": "minutes",
              }
            }
            timeFormat=""
            timeFormatter={null}
            timeMode="24"
            trigger={null}
            useTz={true}
            withoutIcon={false}
          />,
          <span
            className="help-text"
          >
            TimeField: help_text
          </span>,
          <ErrorsList
            color="red"
            header="Errors"
          />,
        ],
        "disabled": true,
        "error": false,
        "required": undefined,
        "width": 1,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "displayOptions": Object {
              "width": 1,
            },
            "help_text": "TimeField: help_text",
            "name": "TimeField: name",
            "type": "TimeField",
            "verbose_name": "TimeField: verbose_name",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoClose": true,
            "autoMode": true,
            "closeOnOutsideClick": true,
            "colorPalette": "light",
            "disabled": false,
            "draggable": true,
            "focused": false,
            "language": "en",
            "limitDrag": false,
            "meridiem": null,
            "minuteStep": 5,
            "onFocusChange": [Function],
            "onTimeChange": [Function],
            "onTimezoneChange": [Function],
            "placeholder": "",
            "showTimezone": true,
            "theme": "material",
            "time": null,
            "timeConfig": Object {
              "step": 30,
              "unit": "minutes",
            },
            "timeFormat": "",
            "timeFormatter": null,
            "timeMode": "24",
            "timezone": undefined,
            "trigger": null,
            "useTz": true,
            "withoutIcon": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "TimeField: help_text",
            "className": "help-text",
          },
          "ref": null,
          "rendered": "TimeField: help_text",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "color": "red",
            "header": "Errors",
            "messages": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
